### PR TITLE
Pub 9

### PR DESCRIPTION
**Pull request by ezeakis**

### CODE REVIEW

File: nttt/utilities.py [hunk 1]
```diff
@@ -53,4 +54,29 @@ def find_files(src, file_names=[], extensions=[]):
...
+    mystring = mystring.replace(" `", "`")
+    return mystring
+
+def find_candidate_strings_for_pub_9(mystring):
+    candidates1 = re.findall(r"_ [a-z]* _", mystring)
+    candidates2 = re.findall(r"` [a-z]* `", mystring)
+    candidates3 = re.findall(r"[*]+ [a-z]* [*]+", mystring)
+    return candidates1 + candidates2 + candidates3
+
+def trim_pub_9(mystring):
```
Comment (robertjank) [inline nttt/utilities.py:78]: Please use function names that describe what the function does. For example, this function could be called trim_special_markdown_tags. Another developer (and probably also you in a couple of days/weeks/months) will not know by heart what pub_9 is, and will have to search in Jira or somewhere else to find out what this function does.

File: nttt/utilities.py [hunk 2]
```diff
@@ -53,4 +54,29 @@ def find_files(src, file_names=[], extensions=[]):
             if valid_file:
                 files_found.append(fpath)
 
-    return files_found
\ No newline at end of file
+    return files_found
+
+
+#PUB-9 utilities
```
Comment (robertjank) [inline nttt/utilities.py:60]: Unless the name of a function and its arguments makes clear what the function does, it is wise to add some documentation above a function, containing the purpose of the function (what it does, not how it does it), a description of its arguments and a description of the return value, if present.
For example, this function name contains the word special. What kind of special thing does this function do?

File: unit_test/test_pub_9.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+import sys
+sys.path.insert(1, '../')
```
Comment (robertjank) [inline unit_test/test_pub_9.py:2]: Why is this necessary?

File: unit_test/test_pub_9.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import unittest
```
Comment (robertjank) [inline unit_test/test_pub_9.py:4]: Please keep the import statements together.

File: unit_test/test_pub_9.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+class TestPub9(unittest.TestCase):
+    def test_pub_9(self):
+    
+        c_initial = 'asd _ fgh _ asd ` ghj ` asd ** hjk ** asd * uio * asd'
+        c_target = 'asd _fgh_ asd `ghj` asd **hjk** asd *uio* asd'
+        
+        self.assertEqual(nttt.utilities.trim_pub_9(c_initial), c_target)
+    
+
```
Comment (robertjank) [inline unit_test/test_pub_9.py:16]: Other cases that could be tested:
c_initial = 'asd _ fgh_ asd ` ghj` asd ** hjk** asd * uio* asd' (only spaces after the open tag)
c_initial = 'asd _fgh _ asd `ghj ` asd **hjk ** asd *uio * asd' (only spaces before the close tag)
c_initial = 'asd _  fgh  _ asd `  ghj  ` asd **  hjk  ** asd *  uio  * asd' (multiple spaces)

File: unit_test/test_pub_9.py (hunk 3 above)
Comment (ezeakis) [reply unit_test/test_pub_9.py:2]: It helps me run the tests locally but you are right I will keep it only for my local tests and remove it from the code

File: unit_test/test_pub_9.py (hunk 5 above)
Comment (ezeakis) [reply unit_test/test_pub_9.py:16]: OK with left and right spaces.

I have problems implementing multiple spaces. There are some complications.

**Comment by pcolladosoto** [comment]
Please check whether this closes issue #12. I haven't taken a thorough look at the pull request but I thought it might be related... Sorry if it isn't though 😅.

File: unit_test/test_pub_9.py (hunk 5 above)
Comment (ezeakis) [reply unit_test/test_pub_9.py:16]: I added some more tests that reveal that there are more problems.
For example
- asd _   fgh _ asd _ ghj   _ asd
+ asd _fgh_ asd _ghj_ asd

It fails by converting it all to 
- asd _fgh_asd_ghj_ asd

The problem here is that underscores are not like eg xml markup where you know which is the beginning and which the ending tag. All underscores are identical for the regular expression. So how can we fix fgh and ghj without affecting asd? Any ideas?

**Comment by ezeakis** [comment]
> Please check whether this closes issue #12. I haven't taken a thorough look at the pull request but I thought it might be related... Sorry if it isn't though 😅.

**Comment by ezeakis** [comment]
If PUB-9 is fixed, then I believe it would be easy to re-use it for issue-12. But at the moment PUB-9 is more challenging than issue-12, I do not know if it is a good idea to merge them.

File: unit_test/test_pub_9.py (hunk 5 above)
Comment (ezeakis) [reply unit_test/test_pub_9.py:16]: Fixed. I added some more tests to be sure.

**Comment by ezeakis** [comment]
@robertjank @wheleph @pcolladosoto 

Please review the code

File: nttt/tidyup.py [hunk 6]
```diff
@@ -1,5 +1,5 @@
 from .utilities import find_files, find_replace, find_snippet, get_file, save_file
-
+from .utilities import trim_spaces_on_specific_markdown
```
Comment (wheleph) [inline nttt/tidyup.py:2]: I would suggest to put this code in a different source file. Something like `cleanup_markup.py`. 

Let's keep only general infrastructure methods for reading, saving, searching etc in `utitlites.py`. And nttt-specific business logic elsewhere.

Comment (wheleph) [inline nttt/tidyup.py:2]: I would suggest to put this code in a different source file. Something like `cleanup_markup.py`.

Let's keep only general infrastructure methods for reading, saving, searching etc in `utitlites.py`. And have nttt-specific business logic elsewhere.

File: nttt/utilities.py [hunk 7]
```diff
@@ -55,4 +57,54 @@ def find_files(src, file_names=[], extensions=[]):
...
-    return files_found
\ No newline at end of file
+    return files_found
+
+
+#PUB-9 utilities. These methods are for trimming spaces on specific markdown like ** or _. The main method is trim_spaces_on_specific_markdown which uses trim_spaces_on_specific_markdown_breakdown and strings_with_specific_markdown_and_spaces.
+ 
+
+regular_expressions_matrix_for_trimming_spaces = [
+    {'character':'_', 'regular_expression_for_wrong_left':r"_[ ]+[a-z]+[ ]*_", 'regular_expression_for_wrong_right':r"_[ ]*[a-z]+[ ]+_", "regular_expression_for_correct":r"_[a-z]*_"},
```
Comment (wheleph) [inline nttt/utilities.py:67]: I think a-z is way too restrictive. It probably should also include capital letters, numbers and underscore (valid character in identifiers).

The original request from Nina also mentions that the problem should be fixed for phrases that contain spaces. On the other hand adding space character in the middle of the regex may lead to some nasty false positives. Consider for example:

> ** is another math operator operator in Python. For example 5 ** 3 = 125

And getting those false positives is even worse than not having the check at all because it may spoil perfect translations.

So I'd suggest to re-evaluate this requirement with Nina and Maja highlighting all possible risks.

File: nttt/utilities.py [hunk 8]
```diff
@@ -55,4 +57,54 @@ def find_files(src, file_names=[], extensions=[]):
...
+
+corrections_for_trimming_spaces = [
+    {'character':'_', 'left_space':"_ ", 'right_space':" _"},
+    {'character':'`', 'left_space':"` ", 'right_space':" `"},
+    {'character':'<asterisk>', 'left_space':"<asterisk> ", 'right_space':" <asterisk>"},
+    {'character':'*', 'left_space':"* ", 'right_space':" *"},
+    ]
+
+def trim_spaces_on_specific_markdown(content):
+    content = content.replace("**", "<asterisk>")
```
Comment (wheleph) [inline nttt/utilities.py:81]: Replacing ** with <asterisk> is misleading because this is for double asterisk.

File: nttt/utilities.py [hunk 9]
```diff
@@ -55,4 +57,54 @@ def find_files(src, file_names=[], extensions=[]):
...
+    for i in [0, 1, 2, 3]:
```
Comment (wheleph) [inline nttt/utilities.py:82]: `range` function would be useful here: https://www.w3schools.com/python/python_for_loops.asp

File: nttt/utilities.py [hunk 10]
```diff
@@ -55,4 +57,54 @@ def find_files(src, file_names=[], extensions=[]):
...
+        #         modified_string = modified_string.replace(corrections_for_trimming_spaces[i]['left_space'], character_on_focus)
+        #         modified_string = modified_string.replace(corrections_for_trimming_spaces[i]['right_space'], character_on_focus)        
+        #         if bool(re.search(regular_expressions_matrix_for_trimming_spaces[i]['regular_expression_for_correct'], modified_string)):
+        #             break
+        #     content = content.replace(problematic_string, modified_string)
+        #     problematic_strings = re.findall(regular_expressions_matrix_for_trimming_spaces[i]['regular_expression_for_wrong'], content)
+            
+        for problematic_string in problematic_strings:
+            modified_string = problematic_string
+            while True:
```
Comment (wheleph) [inline nttt/utilities.py:101]: probably it's possible to replace this entire loop with capture groups (https://regexone.com/lesson/capturing_groups) that allow to fetch the content you're interested in directly.

File: nttt/utilities.py (hunk 7 above)
Comment (wheleph) [reply nttt/utilities.py:67]: Another option would be to introduce feature flags (command line arguments) that enable/disable those "dangerous" features.

**Comment by ezeakis** [comment]
It seems that requirements need to be revisited and the whole way this functionality is create to be reviewed. So this Pull Request is closed.
